From 6a55ca3b0f58b06df098be1cd083b20c9cc6e2ca Mon Sep 17 00:00:00 2001
From: liberodark <liberodark@gmail.com>
Date: Wed, 26 Nov 2025 11:23:24 +0100
Subject: [PATCH 08/18] security: introduce CONFIG_SLAB_VIRTUAL

---
 security/Kconfig.hardening | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/security/Kconfig.hardening b/security/Kconfig.hardening
index c9d5ca3d8..c0062c5e7 100644
--- a/security/Kconfig.hardening
+++ b/security/Kconfig.hardening
@@ -378,4 +378,21 @@ config GCC_PLUGIN_RANDSTRUCT
 	   * https://grsecurity.net/
 	   * https://pax.grsecurity.net/
 
+config SLAB_VIRTUAL
+	bool "Allocate slab objects from virtual memory"
+	depends on SLUB && !SLUB_TINY
+	# If KFENCE support is desired, it could be implemented on top of our
+	# virtual memory allocation facilities
+	depends on !KFENCE
+	# ASAN support will require that shadow memory is allocated
+	# appropriately.
+	depends on !KASAN
+	# SLAB_VIRTUAL is supposed to prevent virtual address reuse between
+	# different caches, merging caches is incompatible with that goal.
+	depends on !SLAB_MERGE_DEFAULT
+	help
+	  Allocate slab objects from kernel-virtual memory, and ensure that
+	  virtual memory used as a slab cache is never reused to store
+	  objects from other slab caches or non-slab data.
+
 endmenu
-- 
2.50.1

