--- a/src/suyu/CMakeLists.txt	2026-04-09 09:48:54.050632549 +0000
+++ b/src/suyu/CMakeLists.txt	2026-04-09 09:49:03.196629552 +0000
@@ -381,14 +381,13 @@
     endif()
 endif()

+find_package(Qt${QT_MAJOR_VERSION} REQUIRED COMPONENTS GuiPrivate)
+
 target_link_libraries(suyu PRIVATE common core input_common frontend_common network video_core)
-target_link_libraries(suyu PRIVATE Boost::headers glad Qt${QT_MAJOR_VERSION}::Widgets)
+target_link_libraries(suyu PRIVATE Boost::headers glad Qt${QT_MAJOR_VERSION}::Widgets Qt${QT_MAJOR_VERSION}::GuiPrivate)
 target_link_libraries(suyu PRIVATE ${PLATFORM_LIBRARIES} Threads::Threads)

 target_link_libraries(suyu PRIVATE Vulkan::Headers)
-if (NOT WIN32)
-    target_include_directories(suyu PRIVATE ${Qt${QT_MAJOR_VERSION}Gui_PRIVATE_INCLUDE_DIRS})
-endif()
 if (UNIX AND NOT APPLE)
     target_link_libraries(suyu PRIVATE Qt${QT_MAJOR_VERSION}::DBus)
 endif()
